### PR TITLE
Prefer Sets#intersection over more contrived alternatives.

### PR DESCRIPTION
**Pull request by mohamedsamehsalah**
~~Original discussion.~~

Suggested commit message:
```
Introduce `Sets{Difference,Intersection}{,Map,Multimap}` and `SetsUnion` Refaster rules (#607)
```

### CODE REVIEW

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ImmutableSetRules.java [hunk 1]
```diff
@@ -211,4 +214,43 @@ ImmutableSet<T> after(T e1, T e2, T e3, T e4, T e5) {
...
+    @AfterTemplate
+    ImmutableSet<K> after(Set<K> set, Map<K, V> map) {
+      return Sets.intersection(set, map.keySet()).immutableCopy();
+    }
+  }
+
+  /** Prefer {@link Sets#intersection(Set, Set)} ()} over more contrived alternatives. */
+  static final class SetMultimapKeySetIntersection<K, V> {
+    @BeforeTemplate
+    ImmutableSet<K> before(Set<K> set, SetMultimap<K, V> setMultimap) {
```
Comment (rickie) [inline error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ImmutableSetRules.java]: We can use a `Multimap` here 😄.

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ImmutableSetRules.java [hunk 2]
```diff
@@ -211,4 +214,43 @@ ImmutableSet<T> after(T e1, T e2, T e3, T e4, T e5) {
       return ImmutableSet.of(e1, e2, e3, e4, e5);
     }
   }
+
+  /** Prefer {@link Sets#intersection(Set, Set)} ()} over more contrived alternatives. */
+  static final class SetsIntersection<T> {
```
Comment (rickie) [inline error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ImmutableSetRules.java:219]: To be honest, it's a bit hard to decide how we should name these rules. We invoke the same method in the three rules but how we get the set that we pass differs. Till now most cases also _use_ different overloads of a method and based on that we use different words.

I imagine we'd also have something that determines a suffix based on what we _pass_ as the two parameters for `Sets#intersection`. 

Was doubting whether we want to add the `ImmutableCopy` as suffix as well 🤔. 

Pushed a suggestion, let's discuss 👀.

**Review by rickie — APPROVED** [review]
Nice improvements @mohamedsamehsalah! 🚀 

They make things more clear 😄.

Added a commit and rebased.

Suggested commit message:
```
Introduce `Sets{Difference,Intersection}{,Map,Multimap}` Refaster rules (#607)
```

File: error-prone-contrib/src/test/resources/tech/picnic/errorprone/refasterrules/ImmutableSetRulesTestInput.java [hunk 3]
```diff
@@ -72,4 +74,21 @@ Set<Integer> testImmutableSetOf4() {
   Set<Integer> testImmutableSetOf5() {
     return Set.of(1, 2, 3, 4, 5);
   }
+
+  ImmutableSet<Integer> testSetsIntersection() {
+    ImmutableSet<Integer> set = ImmutableSet.of(1);
+    return ImmutableSet.of(1, 2, 3).stream().filter(set::contains).collect(toImmutableSet());
```
[the remark this answers is not in the document]
Comment (mohamedsamehsalah) [reply error-prone-contrib/src/test/resources/tech/picnic/errorprone/refasterrules/ImmutableSetRulesTestInput.java]: Overlooked what we are actually testing here 😅 
Thanks 🚀

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ImmutableSetRules.java [hunk 4]
```diff
@@ -211,4 +214,43 @@ ImmutableSet<T> after(T e1, T e2, T e3, T e4, T e5) {
       return ImmutableSet.of(e1, e2, e3, e4, e5);
     }
   }
+
+  /** Prefer {@link Sets#intersection(Set, Set)} over more contrived alternatives. */
+  static final class SetsIntersection<T> {
+    @BeforeTemplate
+    ImmutableSet<T> before(Set<T> set1, Set<T> set2) {
+      return set1.stream().filter(set2::contains).collect(toImmutableSet());
+    }
```
Comment (Stephan202) [inline error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ImmutableSetRules.java]: Here and below: the sets don't need to be of the same type `T` :)

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ImmutableSetRules.java [hunk 5]
```diff
@@ -211,4 +214,43 @@ ImmutableSet<T> after(T e1, T e2, T e3, T e4, T e5) {
...
+
+    @AfterTemplate
+    ImmutableSet<T> after(Set<T> set1, Set<T> set2) {
+      return Sets.intersection(set1, set2).immutableCopy();
```
Comment (Stephan202) [inline error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ImmutableSetRules.java:282]: Next to `intersection` there's also `difference`. (And `union`; see the internal Store PR I filed a few days ago.)

**Review by Stephan202 — COMMENTED** [review]
Added a commit. Did not yet look at `union`; will do another pass later to see whether that's a reasonable addition to this PR.

Nice idea @mohamedsamehsalah!

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ImmutableSetRules.java (hunk 4 above)
Comment (mohamedsamehsalah) [reply error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ImmutableSetRules.java]: Will sets of same type still be refastered after this ?

File: error-prone-contrib/src/test/resources/tech/picnic/errorprone/refasterrules/ImmutableSetRulesTestInput.java [hunk 6]
```diff
@@ -72,4 +74,19 @@ Set<Integer> testImmutableSetOf4() {
...
+    return ImmutableSet.of(2).stream().filter(set::contains).collect(toImmutableSet());
+  }
+
+  ImmutableSet<Integer> testSetsIntersectionMap() {
+    ImmutableMap<Integer, Integer> map = ImmutableMap.of(1, 2);
+    return ImmutableSet.of(3).stream().filter(map::containsKey).collect(toImmutableSet());
+  }
+
+  ImmutableSet<Integer> testSetsIntersectionMultimap() {
+    ImmutableMultimap<Integer, Integer> multimap = ImmutableMultimap.of(1, 2);
```
[the remark this answers is not in the document]
Comment (mohamedsamehsalah) [reply error-prone-contrib/src/test/resources/tech/picnic/errorprone/refasterrules/ImmutableSetRulesTestInput.java]: What is the reason behind this preference ?

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ImmutableSetRules.java (hunk 5 above)
Comment (mohamedsamehsalah) [reply error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ImmutableSetRules.java:282]: For reference: [#2545](https://github.com/PicnicSupermarket/picnic-store/pull/2524/files)

Comment (Stephan202) [reply error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ImmutableSetRules.java:282]: @mohamedsamehsalah we try not to link to private things in this repo; I left it out on purpose. :upside_down_face:

**Comment by mohamedsamehsalah** [comment]
> @Stephan202 
...Nice idea @mohamedsamehsalah!

Original idea from @werli 🚀

File: error-prone-contrib/src/test/resources/tech/picnic/errorprone/refasterrules/ImmutableSetRulesTestInput.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (Stephan202) [reply error-prone-contrib/src/test/resources/tech/picnic/errorprone/refasterrules/ImmutableSetRulesTestInput.java]: Check the [documentation](https://guava.dev/releases/19.0/api/docs/com/google/common/collect/ImmutableMultimap.html) :)

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ImmutableSetRules.java (hunk 4 above)
Comment (Stephan202) [reply error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ImmutableSetRules.java]: yep, it matches for any `S` and `T`, so also of `S == T`.

**Review by Stephan202 — APPROVED** [review]
Added one more commit :)

**Review by werli — APPROVED** [review]
Nice, thanks @mohamedsamehsalah! Also thanks to the other reviewers. I got even more than I suggested back then. :wink:

**Pull request merged**